kimchi/src/circuits/polynomials/keccak/circuitgates.rs

```diff
@@ -130,6 +120,145 @@ where
+//~ |     10 |      `bound_crumb3` | `shifted_crumb3` |
+//~ |     11 |      `bound_crumb4` | `shifted_crumb4` |
+//~ |     12 |      `bound_crumb5` | `shifted_crumb5` |
+//~ |     13 |      `bound_crumb6` | `shifted_crumb6` |
+//~ |     14 |      `bound_crumb7` | `shifted_crumb7` |
+//~
+//~ In Keccak, rotations are performed over a 5x5 matrix state of w-bit words each cell. The values used
+//~ to perform the rotation are fixed, public, and known in advance, according to the following table:
+//~
+//~ | y \ x |   0 |   1 |   2 |   3 |   4 |
+//~ | ----- | --- | --- | --- | --- | --- |
+//~ | 0     |   0 |   1 | 190 |  28 |  91 |
+//~ | 1     |  36 | 300 |   6 |  55 | 276 |
+//~ | 2     |   3 |  10 | 171 | 153 | 231 |
+//~ | 3     | 105 |  45 |  15 |  21 | 136 |
+//~ | 4     | 210 |  66 | 253 | 120 |  78 |
+//~
+//~ But since we are always using 64-bit words, we can have an equivalent table with these values modulo 64
+//~ to avoid needing multiple passes of the rotation gate (a single step would cause overflows):
+//~
+//~ | y \ x |   0 |   1 |   2 |   3 |   4 |
+//~ | ----- | --- | --- | --- | --- | --- |
+//~ | 0     |   0 |   1 |  62 |  28 |  27 |
+//~ | 1     |  36 |  44 |   6 |  55 |  20 |
+//~ | 2     |   3 |  10 |  43 |  25 |  39 |
+//~ | 3     |  41 |  45 |  15 |  21 |   8 |
+//~ | 4     |  18 |   2 |  61 |  56 |  14 |
+//~
+//~ Since there is one value of the coordinates (x, y) where the rotation is 0 bits, we can skip that step in the
+//~ gadget. This will save us one gate, and thus the whole 25-1=24 rotations will be performed in just 48 rows.
+//~
+#[derive(Default)]
+pub struct KeccakRot<F>(PhantomData<F>);
+
+impl<F> Argument<F> for KeccakRot<F>
+where
+    F: PrimeField,
+{
+    const ARGUMENT_TYPE: ArgumentType = ArgumentType::Gate(GateType::KeccakRot);
+    const CONSTRAINTS: u32 = 11;
+
+    // Constraints for rotation of three 64-bit words by any three number of bits modulo 64
+    // (stored in coefficient as a power-of-two form)
+    //   * Operates on Curr row
+    //   * Shifts the words by `rot` bits and then adds the excess to obtain the rotated word.
+    fn constraint_checks<T: ExprOps<F>>(env: &ArgumentEnv<F, T>) -> Vec<T> {
+        // Check that the last 8 columns are 2-bit crumbs
+        let mut constraints = (7..COLUMNS)
+            .map(|i| crumb(&env.witness_curr(i)))
+            .collect::<Vec<T>>();
+
+        let two_to_64 = T::from(2u64).pow(64);
+
+        let word = env.witness_curr(0);
+        let rotated = env.witness_curr(1);
+        let excess = env.witness_curr(2);
+        let shifted = env.witness_next(0);
+        let two_to_rot = env.coeff(0);
+
+        // Obtains the following checks:
+        // word * 2^{rot} = (excess * 2^64 + shifted)
+        // rotated = shifted + excess
+        constraints.push(
+            word * two_to_rot.clone() - (excess.clone() * two_to_64.clone() + shifted.clone()),
+        );
+        constraints.push(rotated - (shifted + excess.clone()));
```

Comment:
Shouldn't this be `rotated = shifted * 2^{64-rot} + excess`?

Reply:
It could, but the way this shifted is defined is to be such that: `word * 2^{rot} = excess * 2^{64} + shifted`, sort of like in foreign field multiplication with quotient and remainders, meaning that this shifted already contains the `2^{64-rot}` term.

Reply:
We could do the other way round if we change the witness generation code accordingly. But since the original idea of this variant of the gate was precisely FFMul, we kept the _quotient & remainder_ apprach.